index.js

```diff
@@ -4,7 +4,7 @@ const { format } = require('util')
 
 function build () {
   const codes = {}
-  const emitted = new Map()
+  const emitted = new Set()
 
   function create (name, code, message) {
```

Comment:
Shouldn't line 16 be updated to something like `if (codes.has(code) === true) {`?

Reply:
> Should I also touch the codes object?

With the proposed change, I do not see the point of the `codes` object.

Reply:
The codes object just stores the warning, as you have to call it once in emit.

Reply:
We could delete used codes. 

```js
'use strict'

const { format } = require('util')

function build () {
  const codes = new Map()
  const emitted = new Set()

  function create (name, code, message) {
    if (!name) throw new Error('Warning name must not be empty')
    if (!code) throw new Error('Warning code must not be empty')
    if (!message) throw new Error('Warning message must not be empty')

    code = code.toUpperCase()

    if (codes.has(code) || emitted.has(code)) {
      throw new Error(`The code '${code}' already exist`)
    }

    function buildWarnOpts (a, b, c) {
      // more performant than spread (...) operator
      let formatted
      if (a && b && c) {
        formatted = format(message, a, b, c)
      } else if (a && b) {
        formatted = format(message, a, b)
      } else if (a) {
        formatted = format(message, a)
      } else {
        formatted = message
      }

      return {
        code,
        name,
        message: formatted
      }
    }

    codes.set(code, buildWarnOpts)

    return buildWarnOpts
  }

  function emit (code, a, b, c) {
    if (emitted.has(code)) {
      return
    }
    if (codes.has(code) === false) {
      throw new Error(`The code '${code}' does not exist`)
    }
    emitted.add(code)

    const warning = codes.get(code)(a, b, c)
    process.emitWarning(warning.message, warning.name, warning.code)
    codes.delete(code)
  }

  return {
    create,
    emit,
    emitted
  }
}

module.exports = build

```

Reply:
replaced codes-Object with Map.

Reply:
I keep forgetting that `Set` is a list of values and not a hash. I think this PR has shown that the current implementation can remove the `codes` plain JavaScript object and utilize a single `Map` for emitted.